### ISSUE TITLE
Access to IMvcBuilder

### ISSUE THREAD

**Issue by rcoode** [OrchardCMS/OrchardCore#2918]
I really need to be able to access the IMvcBuilder returned by AddMvc inside AddOrchardCms in order to configure certain options such as Compatability Version and custom Razor Page routing (using Orchard  Core as a decoupled CMS). Is this currently possible or if not could this be made available possibly by passing to the configure action of the AddOrchardCms method ?

**Comment by jtkech** [comment]
All run in the context of an isolated tenant (we have at least the `Default` tenant), then each module has a startup class with `ConfigureServices()` and `Configure()` methods to configure services options and so on.

The application also behaves as a module and we have helpers (acting on a `OrchardCoreBuilder`) to do configurations in the context of a tenant, as a module. So, through the application you can e.g configure `MvcOptions` and so on. There is an optional order property if you want to run 1st or last, 0 by default.

            services.AddOrchardCms()
                .AddOrchardCore()
                .ConfigureServices(collection =>
                {
                    collection.Configure<MvcOptions>((options) =>
                    {
                    });
                }, 10) // optional order
                .Configure(builder =>
                {
                }, 10); // optional order


About razor page routing,  see the demo module startup which uses one of our specific helper `AddAreaFolderRoute()` and an mvc standard one `AddAreaPageRoute()`. You can do the same in a module startup or through the application by using a `ConfigureServices()`  acting on `OrchardCoreBuilder`.

**Comment by rcoode** [comment]
I tried adding the following code but call the IMvcBuilder extension method **SetCompatabilityVersion** but although it get compiled the line is never run

```
            services.AddOrchardCms()
                .AddOrchardCore()
                .ConfigureServices(coll => coll.Configure<IMvcBuilder>(
                    options => {
                        options.SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
                    }));
```

**Comment by jtkech** [comment]
Yes you're right because it acts on the mvc builder, so try this.

    using Microsoft.AspNetCore.Mvc;
    using Microsoft.AspNetCore.Mvc.Infrastructure;
    ....

            services.AddOrchardCms()
                .AddOrchardCore()
                .ConfigureServices(coll => coll.Configure<MvcCompatibilityOptions>(
                    options => {
                        options.CompatibilityVersion = CompatibilityVersion.Version_2_1;
                    }));

The demo module is in our repo under `src\OrchardCore.Modules\OrchardCore.Demo`, it needs to be referenced by your application if you want to try it.

**Comment by rcoode** [comment]
As it happens my issues were fixed by latest build which I noticed fixed some Razor page routing issues relating to 2.2

Thanks for the help

**Comment by sebastienros** [comment]
We would really appreciate a documentation PR that will help others based on your experience.

**Comment by rcoode** [comment]
Hi Sébastien,

I’m a bit new to open source so I’m not sure what is involved in your request.

We have just completed a re-write of our own web site using Orchard Core as a de-coupled CMS which are currently testing.

https://blcx-website.azurewebsites.net/

The process was very easy and I was impressed with the stability of Orchard Core

On the basis of this experience we have decided to use Orchard for a customer site and I hope to be able to provide feedback to the Orchard Core project as a result of this development project

Kind regards

Robin

**Comment by sebastienros** [comment]
Please let us know when the sites are online, I love to show them during our meetings. And also please share any feedback you have on your experiences.
